File: sdks/node/src/serialize.ts
import * as protobuf from 'protobufjs';
import * as path from 'path';
import { generateULID } from './ulid';

let ProtoRoot: protobuf.Root | null = null;
let LogEventType: protobuf.Type | null = null;
let IngestBatchType: protobuf.Type | null = null;
let IngestResponseType: protobuf.Type | null = null;

export interface LogEventData {
  service_name: string;
  message: string;
  level: string;
  timestamp_ns: number;
  logger_name?: string;
  file?: string;
  line?: number;
  correlation_id?: string;
  attributes?: Record<string, string>;
  stack_trace?: string[];
  exception_type?: string;
  exception_message?: string;
  event_id?: string;
}

export interface IngestBatchData {
  service_name: string;
  instance_id: string;
  batch_seq: number;
  logs: LogEventData[];
  spans: any[];
  metrics: any[];
}

export function initProtobuf(protoPath?: string): void {
  if (ProtoRoot) return;

  const resolvedPath = protoPath ?? path.join(__dirname, '..', 'proto', 'greplog', 'v1', 'events.proto');
  ProtoRoot = protobuf.loadSync(resolvedPath);
  LogEventType = ProtoRoot.lookupType('greplog.v1.LogEvent');
  IngestBatchType = ProtoRoot.lookupType('greplog.v1.IngestBatch');
  IngestResponseType = ProtoRoot.lookupType('greplog.v1.IngestResponse');
}

/**
 * Convert a LogEventData (snake_case) to the camelCase keys that
 * protobufjs create() requires. protobufjs silently drops snake_case
 * keys — every multi-word field (event_id, timestamp_ns, etc.) must
 * be camelCased or it won't appear on the wire.
 */
function logEventToCamelCase(data: LogEventData): Record<string, any> {
  const payload: Record<string, any> = {
    serviceName: data.service_name,
    message: data.message,
    level: data.level,
    timestampNs: data.timestamp_ns,
    eventId: data.event_id || generateULID(),
  };

  if (data.logger_name) payload.loggerName = data.logger_name;
  if (data.file) payload.file = data.file;
  if (data.line !== undefined) payload.line = data.line;
  if (data.correlation_id) payload.correlationId = data.correlation_id;
  if (data.attributes && Object.keys(data.attributes).length > 0) {
    payload.attributes = data.attributes;
  }
  if (data.stack_trace && data.stack_trace.length > 0) {
    payload.stackTrace = data.stack_trace;
  }
  if (data.exception_type) payload.exceptionType = data.exception_type;
  if (data.exception_message) payload.exceptionMessage = data.exception_message;

  return payload;
}

export function encodeLogEvent(data: LogEventData): Uint8Array {
  if (!LogEventType) initProtobuf();

  const payload = logEventToCamelCase(data);
  const message = LogEventType!.create(payload);
  return LogEventType!.encode(message).finish() as Uint8Array;
}

export function encodeIngestBatch(data: IngestBatchData): Uint8Array {
  if (!IngestBatchType) initProtobuf();

  const logs = data.logs.map((log) => logEventToCamelCase(log));

  const message = IngestBatchType!.create({
    serviceName: data.service_name,
    instanceId: data.instance_id,
    batchSeq: data.batch_seq,
    logs,
    spans: data.spans ?? [],
    metrics: data.metrics ?? [],
  });

  return IngestBatchType!.encode(message).finish() as Uint8Array;
}

export function decodeIngestResponse(buf: Uint8Array): { accepted: boolean; events_count: number; error: string } | null {
  if (!IngestResponseType) initProtobuf();
  try {
    const msg = IngestResponseType!.decode(buf);
    const obj = IngestResponseType!.toObject(msg, { defaults: true }) as any;
    return { accepted: !!obj.accepted, events_count: obj.eventsCount ?? 0, error: obj.error ?? '' };
  } catch {
    return null;
  }
}

export function encodeLengthPrefixedFrame(payload: Uint8Array): Buffer {
  const len = Buffer.alloc(4);
  len.writeUInt32LE(payload.length);
  return Buffer.concat([len, Buffer.from(payload)]);
}

File: sdks/python/greplog/serialize.py
"""Protobuf serialization for IngestBatch wire frames."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from greplog.proto.greplog.v1 import events_pb2
from greplog.ulid import generate_ulid


def encode_length_prefixed_frame(payload: bytes) -> bytes:
    return len(payload).to_bytes(4, byteorder="little") + payload


def _build_log_event(data: Dict[str, Any]) -> events_pb2.LogEvent:
    event = events_pb2.LogEvent(
        service_name=data.get("service_name", ""),
        message=data.get("message", ""),
        level=data.get("level", ""),
        timestamp_ns=int(data.get("timestamp_ns", 0)),
        event_id=data.get("event_id") or generate_ulid(),
    )

    if data.get("logger_name"):
        event.logger_name = data["logger_name"]
    if data.get("file"):
        event.file = data["file"]
    if data.get("line") is not None:
        event.line = int(data["line"])
    if data.get("correlation_id"):
        event.correlation_id = data["correlation_id"]
    if data.get("attributes"):
        for key, val in data["attributes"].items():
            event.attributes[key] = str(val)
    if data.get("stack_trace"):
        event.stack_trace.extend(data["stack_trace"])
    if data.get("exception_type"):
        event.exception_type = data["exception_type"]
    if data.get("exception_message"):
        event.exception_message = data["exception_message"]

    return event


def encode_ingest_batch(
    service_name: str,
    instance_id: str,
    batch_seq: int,
    logs: List[Dict[str, Any]],
    spans: Optional[List[Any]] = None,
    metrics: Optional[List[Any]] = None,
) -> bytes:
    batch = events_pb2.IngestBatch(
        service_name=service_name,
        instance_id=instance_id,
        batch_seq=batch_seq,
    )
    batch.logs.extend(_build_log_event(log) for log in logs)
    # spans and metrics intentionally empty for now
    _ = spans, metrics
    return batch.SerializeToString()
